generic/tests/ping.py

```diff
@@ -1,133 +1,182 @@
+            test.fail("Ping failed, status: %s,"
+                      " output: %s" % (status, output))
 
     timeout = int(params.get("login_timeout", 360))
     ping_ext_host = params.get("ping_ext_host", "no") == "yes"
 
     vm = env.get_vm(params["main_vm"])
     vm.verify_alive()
-    error.context("Login to guest", logging.info)
+    error_context.context("Login to guest", logging.info)
     session = vm.wait_for_login(timeout=timeout)
 
     # most of linux distribution don't add IP configuration for extra nics,
     # so get IP for extra nics via pre_cmd;
     if params.get("pre_cmd"):
         session.cmd(params["pre_cmd"], timeout=600)
 
+    ext_host = "www.redhat.com"
     if ping_ext_host:
-        default_host = "www.redhat.com"
         ext_host_get_cmd = params.get("ext_host_get_cmd", "")
         try:
             ext_host = utils.system_output(ext_host_get_cmd)
         except error.CmdError:
             logging.warn("Can't get specified host with cmd '%s',"
                          " Fallback to default host '%s'",
-                         ext_host_get_cmd, default_host)
-            ext_host = default_host
-
-        if not ext_host:
-            # Fallback to a hardcode host, eg:
-            ext_host = default_host
+                         ext_host_get_cmd, ext_host)
 
-    counts = params.get("ping_counts", 100)
+    counts = params.get("counts", 10)
     flood_minutes = float(params.get("flood_minutes", 10))
 
     packet_sizes = params.get("packet_size", "").split()
-
-    for i, nic in enumerate(vm.virtnet):
-        ip = vm.get_address(i)
-        if ip.upper().startswith("FE80"):
-            interface = utils_net.get_neigh_attch_interface(ip)
-        else:
-            interface = None
-        nic_name = nic.get("nic_name")
-        if not ip:
-            logging.error("Could not get the ip of nic index %d: %s",
+    interval_times = params.get("interval_time", "").split()
+    host_result_str = params.get("host_exceeded_res", "")
+    result_str = params.get("guest_exceeded_res", "")
+
+    def _ping_with_params(dest=ext_host, interface=None,
+                          packet_size=None, interval=None,
+                          count=counts, session=None,
+                          flood=False, timeout=float(counts) * 1.5):
+        if flood:
+            count = None
+        return utils_test.ping(dest, count, interval, interface,
+                               packet_size, session=session,
+                               flood=flood, timeout=timeout)
+
+    if params.get("ping_default", "no") == "yes":
+        for i, nic in enumerate(vm.virtnet):
+            ip = vm.get_address(i)
+            if ip.upper().startswith("FE80"):
+                interface = utils_net.get_neigh_attch_interface(ip)
+            else:
+                interface = None
+            nic_name = nic.get("nic_name")
+            if not ip:
+                test.fail("Could not get the ip of nic index %d: %s",
```

Comment:
The previous logic is to pick up one valid ip address (and its attached iface), so do we have some reason to change to the current logic?

Reply:
The previous logic does not have 'if params.get("ping_default", "no") == "yes":' , other codes are the same. If we run ping.ext_host, will also run the ping_default part with previous codes.

For this part, it means we get all the guests' IPs and ping them, it is used for "ping.multi_nics" , not only pick one IP.

Reply:
Ok, got the point, but there is no guarantee that every vm's nic should have an ip address without properly configured (e.g. an additional nic without ifcfg configuration file of a linux guest).

Reply:
In line52 and line53, it get a pre_cmd in cfg (pre_cmd = "for nic in ls /sys/class/net|grep -v lo;do arp -a|grep -v $nic && dhclient $nic;done"), it can dhclient ip for multi_nics. and then judge the ip in line91 is enough. If add a ifcfg, it maybe the same function as line91.